### PR TITLE
test(server): improve e2e coverage for bid selection, relay failure, and invalid payloads

### PR DESCRIPTION
**Pull request by AnkushinDaniil**
## 📝 Summary

This PR introduces end-to-end tests for critical scenarios in MEV-Boost's validator-relay interaction model. The changes target previously uncovered edge cases in the bid - payload request lifecycle, particularly in situations involving multiple relays, malformed payloads, and mismatches in slot

> [!NOTE]
> This PR may be easier to review per commit.

## ⛱ Motivation and Context

This PR was performed in the context of test work.

### Development Notes – Test Strategy & Thought Process

At the beginning of this task, I started by reviewing the existing test coverage in the `mev-boost` repository. I ran the coverage tools locally and considered extending test cases across different fork versions (Bellatrix, Capella, Deneb, Electra). However, after a closer look, I realised that doing so would result in relatively shallow or repetitive tests — and wouldn’t constitute a "meaningful improvement" as requested by the task. The logic across forks is largely duplicated and not where current gaps lie.

Next, I explored the idea of setting up a CI-based integration test pipeline using [`builder-playground`](https://github.com/flashbots/builder-playground). My goal was to create a test harness simulating full validator-relay-builder interaction flows. However, I quickly discovered that:

- Integration tests already exist,
- They use Kurtosis, which is explicitly noted in the task brief as not recommended due to slow iteration time,
- And setting up a reliable CI harness using this stack would far exceed the 10-hour time budget.

At that point, I changed direction and shifted my focus entirely to reading and analysing every test in the server package. This deep dive surfaced something far more impactful:

Certain real-world failure modes — particularly around bid cache handling, invalid payloads, and signature failures — were not covered, or only implicitly assumed in logic but not tested directly.

This observation shaped my final test strategy.

Rather than adding shallow fork permutations or overengineering infra, I chose to focus on high-value, realistic validator/relay edge cases that:

- Are critical to production safety,
- Reflect MEV-Boost's real role as a stateless bid router,
- And are currently untested in the local e2e test suite.

Each new test was designed to explicitly check behavior that would otherwise silently fail.

#### What it tests:

- Bid selection from multiple relays with varying response times.
- Whether MEV-Boost correctly selects the highest bid, even if it arrives later than a lower one.
- Behavior when the selected (winning) relay fails before payload delivery.

#### Why it matters:

- This simulates a very realistic scenario: multiple relays race to offer bids, the validator waits to pick the most valuable one, and the relay that wins might go offline by the time the payload is requested.

#### What I learned while writing this:
Initially, I assumed MEV-Boost might retry another relay if the winning one fails. It does not — and this is by design. Bids are matched one-to-one for security and simplicity. If the chosen relay fails, Boost logs the error and returns a 502. This test proves that.

### `TestSemanticallyInvalidSignedBlindedBlock`

1. `TamperExecutionPayloadHeader`

#### What it tests:

- What happens when a validator tampers with the `ExecutionPayloadHeader` after receiving the header from the relay (e.g., changes `blockHash`).

#### Why it matters:

- In a malicious or misconfigured setup, a validator might attempt to fake a signed payload that doesn’t match the originally agreed bid.
- This is explicitly rejected by MEV-Boost: it can’t find a bid for the new blockHash → returns 502.

#### Key takeaway:

- MEV-Boost doesn’t “know” whether the payload is valid in the cryptographic sense.

It’s a stateless cache match — if `blockHash` doesn't match any stored bid → it must reject.

2. `EmptySignature`

#### What it tests:

- A validator sends a block with an empty BLS signature.

#### Why it matters:

- This case is subtly important. While MEV-Boost itself doesn’t validate BLS signatures, the relays do.
- If the relay rejects the request with a 400 Bad Request, Boost should detect that no relays succeeded and propagate a 502.

#### TODO:

- Consider improving error handling, i.e. return 400 Bad Request in this case.

#### What it tests:

- Submitting a signed blinded block for slot N+1 after getting a bid for slot N.

#### Why it matters:

- In reality, validators might get delayed (clock drift, GC, load spike) and send the signed block slightly late.
- This test confirms that MEV-Boost does not perform any kind of match — slot numbers must match exactly.

## 📚 References

- https://docs.flashbots.net/flashbots-mev-boost/architecture-overview/block-proposal
- https://github.com/flashbots/mev-boost-relay/blob/fdb359fa6b6a7f96d37fb1f8cabb02c3868f965f/services/api/service.go#L654-L660

---

## ✅ I have run these commands

* [x] `make lint`
* [x] `make test-race`
* [x] `go mod tidy`

### CODE REVIEW

File: server/e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+		})
+		// Validate the signature
+		// https://github.com/flashbots/mev-boost-relay/blob/fdb359fa6b6a7f96d37fb1f8cabb02c3868f965f/services/api/service.go#L654-L660
+		relay.OverrideHandleGetPayload(func(w http.ResponseWriter, _ *http.Request) {
+			w.Header().Set("Content-Type", MediaTypeJSON)
+			w.WriteHeader(http.StatusBadRequest)
+			if _, err := w.Write([]byte(`{"code": 400, "message": "could not verify payload signature"}`)); err != nil {
+				t.Fatalf("failed to write response: %v", err)
+			}
+		})
```
Comment (jtraglia) [inline server/e2e_test.go]: Why do we call `OverrideHandleGetPayload` twice?

File: server/e2e_test.go
Comment (jtraglia) [inline server/e2e_test.go:1]: I feel like a valid/success e2e test case would be useful here. That's missing, right?

File: server/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+			},
+		},
+	}
+}
+
+func closeServers(relays []*mock.Relay) {
+	for _, relay := range relays {
+		relay.Server.Close()
+	}
+}
```
Comment (jtraglia) [inline server/e2e_test.go]: Given that this is only used once, I think it would best to move it into `setupTestBackend`. Unless there's a reason this needs to be a separate function.

File: server/e2e_test.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	relay.GetHeaderResponse = relay.MakeGetHeaderResponse(
+		10_000_000_000,
+		parentHash.String(),
+		parentHash.String(),
+		testPubKey,
+		spec.DataVersionElectra,
+	)
+
+	bid := setupBasicRequest(t, backend)
+	signed := createSignedBlindedBlock(bid.Electra.Message.Header, parentHash, slotMismatch, testSig)
```
Comment (jtraglia) [inline server/e2e_test.go]: This variable name is a bit ambiguous. I think `signedBlindedBlock` would be better.

File: server/e2e_test.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+		slowRelay.GetPayloadResponse = blindedBlockToBlockResponse(signed)
+		slowRelay.Server.Close()
+
+		reqHeader := http.Header{}
+		reqHeader.Set("Content-Type", "application/json")
+		resp := backend.request(t, http.MethodPost, params.PathGetPayload, reqHeader, signed)
+
+		require.Equal(t, http.StatusBadGateway, resp.Code)
+		require.Contains(t, strings.ToLower(resp.Body.String()), "no successful relay response")
+	})
```
Comment (jtraglia) [inline server/e2e_test.go]: What is the builder failure here exactly?

**Review by jtraglia — COMMENTED** [review]
Overall, this is pretty decent. The organization is good. I would have appreciated some more comments, and a success case as a sanity test. In a true e2e test, I think it would be best to use the actual relay (maybe via kurtosis). But given the time constraints and all, this makes sense 🙂

File: server/e2e_test.go (hunk 1 above)
Comment (AnkushinDaniil) [reply server/e2e_test.go]: Good catch! In an earlier version of these tests, there were other test cases, and I unintentionally left one of them in.
[test(server): remove duplicate relay override in test](https://github.com/AnkushinDaniil/mev-boost/pull/1/commits/86762348e1f6c955b0e9d460404df1b037e3b94e)

File: server/e2e_test.go
Comment (AnkushinDaniil) [reply server/e2e_test.go:1]: Yes. I decided it didn't need to be in this particular test because I felt like other tests covered it well enough
[test(server): add successful signed blinded block test](https://github.com/AnkushinDaniil/mev-boost/pull/1/commits/2803acc78019749c2967b5d9b3d04f6afe89fd62)

File: server/e2e_test.go (hunk 2 above)
Comment (AnkushinDaniil) [reply server/e2e_test.go]: Oh, you're absolutely right! This function is the result of a bunch of refactoring, and I used it in a number of places before
[refactor(test): inline closeServers function](https://github.com/AnkushinDaniil/mev-boost/pull/1/commits/043a16a853391f4294224a90dafdb04a049e728e)

File: server/e2e_test.go (hunk 3 above)
Comment (AnkushinDaniil) [reply server/e2e_test.go]: I 100% agree 👌
[refactor(test): rename signed to signedBlindedBlock](https://github.com/AnkushinDaniil/mev-boost/pull/1/commits/a0eeeba876ce7f140c59164dc23f194eba450f85)

File: server/e2e_test.go (hunk 4 above)
Comment (AnkushinDaniil) [reply server/e2e_test.go]: My bad, I should've added more descriptive comments
[docs(test): add explanatory comments to e2e tests](https://github.com/AnkushinDaniil/mev-boost/pull/1/commits/a4c01fa05ef0bab06d0613ebb1ea3f7c7cc6565d)

Comment (jtraglia) [reply server/e2e_test.go]: Thanks. Ah I see the `slowRelay.Server.Close()` now.

**Review by jtraglia — COMMENTED** [review]
Thanks for making those changes. This looks better now 🙂
